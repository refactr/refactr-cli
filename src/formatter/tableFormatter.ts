import isArray from 'lodash/isArray';
import Table from 'cli-table';

import humanify from './humanify';
import { Formatter } from './formatter';

type Row = { [key: string]: any };
type Input = Row | Row[];

const tableFormatter: Formatter = function tableFormatter(
  input: Input,
  fields: string[]
): string {
  if (!isArray(input)) {
    input = [input];
  }

  const rows = (input as Row[]).map((row) =>
    fields.reduce<any[]>((acc, field) => {
      const value = row[field];
      acc.push(humanify(value));

      return acc;
    }, [])
  );

  const table = new Table({ head: fields });

  table.push(...rows);

  return table.toString();
};

export default tableFormatter;
