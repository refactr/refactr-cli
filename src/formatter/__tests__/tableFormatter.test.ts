import tableFormatter from '../tableFormatter';

describe('tableFormatter', () => {
  test('formats data as ASCII table', () => {
    // Array
    expect(
      tableFormatter(
        [
          {
            a: 1,
            b: 2,
            c: 3,
          },
          { a: 4, b: 5, c: 6 },
        ],
        ['a', 'b', 'c']
      )
    ).toMatchSnapshot();

    // Single row
    expect(
      tableFormatter(
        {
          a: 1,
          b: 2,
          c: 3,
        },
        ['a', 'b', 'c']
      )
    ).toMatchSnapshot();
  });

  test('omits data fields if not presented in `fields` argument', () => {
    expect(
      tableFormatter([{ foo: 'foo', bar: 'bar', baz: 'baz' }], ['foo', 'bar'])
    ).toMatchSnapshot();
  });
});
