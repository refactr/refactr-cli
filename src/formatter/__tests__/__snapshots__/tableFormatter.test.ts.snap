// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tableFormatter formats data as ASCII table 1`] = `
"[37m┌───┬───┬───┐[39m
[37m│[39m[31m a [39m[37m│[39m[31m b [39m[37m│[39m[31m c [39m[37m│[39m
[37m├───┼───┼───┤[39m
[37m│[39m 1 [37m│[39m 2 [37m│[39m 3 [37m│[39m
[37m├───┼───┼───┤[39m
[37m│[39m 4 [37m│[39m 5 [37m│[39m 6 [37m│[39m
[37m└───┴───┴───┘[39m"
`;

exports[`tableFormatter formats data as ASCII table 2`] = `
"[37m┌───┬───┬───┐[39m
[37m│[39m[31m a [39m[37m│[39m[31m b [39m[37m│[39m[31m c [39m[37m│[39m
[37m├───┼───┼───┤[39m
[37m│[39m 1 [37m│[39m 2 [37m│[39m 3 [37m│[39m
[37m└───┴───┴───┘[39m"
`;

exports[`tableFormatter omits data fields if not presented in \`fields\` argument 1`] = `
"[37m┌─────┬─────┐[39m
[37m│[39m[31m foo [39m[37m│[39m[31m bar [39m[37m│[39m
[37m├─────┼─────┤[39m
[37m│[39m foo [37m│[39m bar [37m│[39m
[37m└─────┴─────┘[39m"
`;
